### PR TITLE
add limited metadata caching to journald and other journal improvements

### PR DESCRIPTION
**Pull request by poettering**

### CODE REVIEW

File: src/basic/string-util.h [hunk 1]
```diff
@@ -200,3 +200,10 @@ DEFINE_TRIVIAL_CLEANUP_FUNC(char *, string_free_erase);
 #define _cleanup_string_free_erase_ _cleanup_(string_free_erasep)
 
 bool string_is_safe(const char *p) _pure_;
+
+static inline size_t strlen_ptr(const char *s) {
```
Comment (vcaputo) [inline src/basic/string-util.h:204]: this seems quite unrelated to the topic of this commit

File: src/journal/journald-context.c [hunk 2]
```diff
@@ -0,0 +1,626 @@
...
+#include "journald-context.h"
+#include "process-util.h"
+#include "user-util.h"
+#include "audit-util.h"
+#include "string-util.h"
+#include "cgroup-util.h"
+
+/* This implements a metadata cache for clients, which are identified by their PID. Requesting metadata through /proc
+ * is expensive, hence let's cache the data if we can. Note that this means the metadata might be out-of-date when we
+ * store it, but it might already be anyway, as we request the data asynchronously from /proc at a different time the
```
Comment (vcaputo) [inline src/journal/journald-context.c]: nit: the current situation is actually the opposite of out of date; the metadata is sampled in the future relative to when the log was produced.

File: src/journal/journald-context.c [hunk 3]
```diff
@@ -0,0 +1,626 @@
...
+
+        r = prioq_ensure_allocated(&s->client_contexts_lru, client_context_compare);
+        if (r < 0)
+                return r;
+
+        c = new0(ClientContext, 1);
+        if (!c)
+                return -ENOMEM;
+
+        c->n_ref = 0;
```
Comment (vcaputo) [inline src/journal/journald-context.c]: redundant to new0()

File: src/journal/journald-context.c [hunk 4]
```diff
@@ -0,0 +1,626 @@
...
+                c->uid = ucred->uid;
+        else
+                (void) get_process_uid(c->pid, &c->uid);
+
+        if (ucred && gid_is_valid(ucred->gid))
+                c->gid = ucred->gid;
+        else
+                (void) get_process_gid(c->pid, &c->gid);
+
+        return 0;
```
Comment (vcaputo) [inline src/journal/journald-context.c]: why isn't this a void function if the errors are all being ignored?

File: src/journal/journald-context.c [hunk 5]
```diff
@@ -0,0 +1,626 @@
...
+}
+
+static int client_context_read_basic(ClientContext *c) {
+        char *t;
+
+        assert(c);
+        assert(pid_is_valid(c->pid));
+
+        if (get_process_comm(c->pid, &t) >= 0) {
```
Comment (vcaputo) [inline src/journal/journald-context.c]: Why not just use the `free_and_replace()` macro from `basic/alloc-util.h`?  These can all lose the {}s even then.

File: src/journal/journald-context.c [hunk 6]
```diff
@@ -0,0 +1,626 @@
...
+                free(c->cmdline);
+                c->cmdline = t;
+        }
+
+        if (get_process_capeff(c->pid, &t) >= 0) {
+                free(c->capeff);
+                c->capeff = t;
+        }
+
+        return 0;
```
Comment (vcaputo) [inline src/journal/journald-context.c]: again; why not a void return?

File: src/journal/journald-context.c [hunk 7]
```diff
@@ -0,0 +1,626 @@
...
+        if (label_size > 0) {
+                char *l;
+
+                /* If we got an SELinux label passed in it counts. */
+
+                l = new(char, label_size + 1);
+                if (!l)
+                        return -ENOMEM;
+
+                memcpy(l, label, label_size);
```
Comment (vcaputo) [inline src/journal/journald-context.c]: is there really not a convenience helper in systemd for doing this null-terminated memcpy?

File: src/journal/journald-context.c [hunk 8]
```diff
@@ -0,0 +1,626 @@
...
+                l[label_size] = 0;
+
+                free(c->label);
```
Comment (vcaputo) [inline src/journal/journald-context.c]: `free_and_replace()`

File: src/journal/journald-context.c [hunk 9]
```diff
@@ -0,0 +1,626 @@
...
+                c->label_size = label_size;
+        }
+#ifdef HAVE_SELINUX
+        else {
+                char *con;
+
+                /* If we got no SELinux label passed in, let's try to acquire one */
+
+                if (getpidcon(c->pid, &con) >= 0) {
+                        free(c->label);
```
Comment (vcaputo) [inline src/journal/journald-context.c]: `free_and_replace()`

File: src/journal/journald-context.c [hunk 10]
```diff
@@ -0,0 +1,626 @@
...
+                return 0;
+        }
+
+        /* Let's shortcut this if the cgroup path didn't change */
+        if (streq_ptr(c->cgroup, t)) {
+                free(t);
+                return 0;
+        }
+
+        free(c->cgroup);
```
Comment (vcaputo) [inline src/journal/journald-context.c]: lots of `free_and_replace()` in this function, I'll stop pointing them out.

File: src/journal/journald-context.c [hunk 11]
```diff
@@ -0,0 +1,626 @@
...
+        (void) client_context_read_invocation_id(s, c);
+
+        c->timestamp = timestamp;
+
+        if (c->in_lru) {
+                assert(c->n_ref == 0);
+                assert_se(prioq_reshuffle(s->client_contexts_lru, c, &c->lru_index) >= 0);
+        }
+
+        return 0;
```
Comment (vcaputo) [inline src/journal/journald-context.c]: why not void?

File: src/journal/journald-context.c [hunk 12]
```diff
@@ -0,0 +1,626 @@
...
+        if (ucred && uid_is_valid(ucred->uid) && c->uid != ucred->uid)
+                goto refresh;
+
+        if (ucred && gid_is_valid(ucred->gid) && c->gid != ucred->gid)
+                goto refresh;
+
+        if (label_size > 0 && (label_size != c->label_size || memcmp(label, c->label, label_size) != 0))
+                goto refresh;
+
+        return 0;
```
Comment (vcaputo) [inline src/journal/journald-context.c]: another void return no?  these all imply errors are propagated and that's simply not the case

File: src/journal/journald-context.c [hunk 13]
```diff
@@ -0,0 +1,626 @@
...
+
+refresh:
+        return client_context_really_refresh(s, c, ucred, label, label_size, unit_id, timestamp);
+}
+
+static void client_context_make_room(Server *s, size_t limit) {
```
Comment (vcaputo) [inline src/journal/journald-context.c]: This function name makes me expect the argument to be how much room to make.

I think it would be better named something like `client_context_try_shrink_to()` or something.

File: src/journal/journald-context.c [hunk 14]
```diff
@@ -0,0 +1,626 @@
...
+        client_context_make_room(s, CACHE_MAX-1);
+
+        r = client_context_new(s, pid, &c);
+        if (r < 0)
+                return r;
+
+        if (add_ref)
+                c->n_ref = 1;
+        else {
+                c->n_ref = 0;
```
Comment (vcaputo) [inline src/journal/journald-context.c]: n_ref is already 0 `from client_context_new()`

File: src/journal/journald-context.c [hunk 15]
```diff
@@ -0,0 +1,626 @@
...
+
+        if (!s->pid1_context) {
+
+                r = client_context_acquire(s, 1, NULL, NULL, 0, NULL, &s->pid1_context);
+                if (r < 0)
+                        log_warning_errno(r, "Failed to acquire PID1's context, ignoring: %m");
+
+        }
+
+        return 0;
```
Comment (vcaputo) [inline src/journal/journald-context.c]: why not a void return?

**Comment by vcaputo** [comment]
Neat, something in this vein is long overdue.

I just did a pretty casual review, mostly nits, otherwise 👍 

It's good you addressed all the tiny allocation/frees for the various metadata fields.  When I started reviewing I was concerned you wouldn't address that aspect while adding the cache.  Your alloca approach is less invasive than mine was, though I avoided all copies.  I think I like yours more.

**Comment by poettering** [comment]
Thanks for the review! I have now force pushed a new version with almost all of your points fixed. I did leave some functions returning "int", even though the caller ignores it then. It just feels weird to eat obvious OOM issues right away in the callee, it felt more natural to leave this to the caller. I mean, ultimately it doesn't really matter anyway, the compiler should optimize all this away easily as this stuff is all static, non-exported stuff...

Anway, I hope that makes some sense. Please have another look so that we can get this landed!

File: src/journal/journald-rate-limit.c [hunk 16]
```diff
@@ -216,6 +216,13 @@ int journal_rate_limit_test(JournalRateLimit *r, const char *id, int priority, u
 
         assert(id);
 
+        /* Returns:
+         *
+         * 0     → the log message shall be suppressed,
+         * 1 + n → if the log message shall be permitted, and n messages where dropped from the peer before
```
Comment (keszybz) [inline src/journal/journald-rate-limit.c]: <strike>if</strike>, and where → were

File: src/basic/process-util.c [hunk 17]
```diff
@@ -388,7 +388,7 @@ int is_kernel_thread(pid_t pid) {
         bool eof;
         FILE *f;
 
-        if (pid == 0 || pid == 1) /* pid 1, and we ourselves certainly aren't a kernel thread */
+        if (pid == 0 || pid == 1 || pid == getpid()) /* pid 1, and we ourselves certainly aren't a kernel thread */
```
Comment (keszybz) [inline src/basic/process-util.c]: Shouldn't this be getpid_cached()? Without that, this additional check might slow things down.

File: src/basic/string-util.h (hunk 1 above)
Comment (keszybz) [reply src/basic/string-util.h:204]: Seems OK to me.

File: src/basic/alloc-util.h [hunk 18]
```diff
@@ -53,6 +55,19 @@ static inline void *mfree(void *memory) {
 
 void* memdup(const void *p, size_t l) _alloc_(2);
 
+_alloc_(2) static inline void *memdup_suffix0(const void *p, size_t l) {
+        void *q;
+
+        /* The same as memdup() but place a safety NUL byte after the allocated memory */
+
+        q = memdup(p, l+1);
```
Comment (keszybz) [inline src/basic/alloc-util.h]: This doesn't look right. Based on the commit description, this could be used to bytes from a fixed size buffer, right to the edge. Then this `memdup` will read one past the allowed area. This must be replaced by `malloc` + `memcpy`.

File: src/journal/journald-context.c [hunk 19]
```diff
@@ -0,0 +1,600 @@
...
+
+  You should have received a copy of the GNU Lesser General Public License
+  along with systemd; If not, see <http://www.gnu.org/licenses/>.
+***/
+
+#ifdef HAVE_SELINUX
+#include <selinux/selinux.h>
+#endif
+
+#include <assert.h>
```
Comment (keszybz) [inline src/journal/journald-context.c]: Don't include assert.h. We redefine `assert` independently.

File: src/journal/journald-context.c [hunk 20]
```diff
@@ -0,0 +1,600 @@
...
+
+#include "alloc-util.h"
+#include "journald-context.h"
+#include "process-util.h"
+#include "user-util.h"
+#include "audit-util.h"
+#include "string-util.h"
+#include "cgroup-util.h"
```
Comment (keszybz) [inline src/journal/journald-context.c]: Any particular reason to have those unsorted?

File: src/journal/journald-context.c (hunk 2 above)
Comment (keszybz) [reply src/journal/journald-context.c]: Yeah, that's complicated. We might actually have accurate metadata more often after this patch, e.g. if a client logs twice, and exits immediately after, and we manage to cache metadata after the first log entry.

File: src/journal/journald-server.c [hunk 21]
```diff
@@ -1147,45 +927,21 @@ void server_dispatch_message(
...
-                c = strchr(c+1, '/');
-                if (c) {
-                        c = strchr(c+1, '/');
-                        if (c)
-                                *c = 0;
-                }
+                /* Write a suppression message if we suppressed something */
+                if (rl > 1)
+                        server_driver_message(s, "MESSAGE_ID=" SD_MESSAGE_JOURNAL_DROPPED_STR,
+                                              LOG_MESSAGE("Suppressed %u messages from %s", rl - 1, path),
```
Comment (glasser) [inline src/journal/journald-server.c]: Won't `path` always be NULL here?

File: src/journal/journald-context.c [hunk 22]
```diff
@@ -0,0 +1,600 @@
...
+
+        assert(c);
+
+        /* Try to acquire the current cgroup path */
+        r = cg_pid_get_path_shifted(c->pid, s->cgroup_root, &t);
+        if (r < 0) {
+
+                /* If that didn't work, we use the unit ID passed in as fallback, if we have nothing cached yet */
+                if (unit_id && !c->unit) {
+                        c->unit = strdup(unit_id);
```
Comment (keszybz) [inline src/journal/journald-context.c:258]: Hmm, is it on purpose that in `client_context_read_label` `free_and_replace` is used, and here just normal assignment (no `free`)?

File: src/journal/journald-context.c [hunk 23]
```diff
@@ -0,0 +1,600 @@
...
+                free(t);
+                return 0;
+        }
+
+        free_and_replace(c->cgroup, t);
+
+        if (cg_path_get_session(c->cgroup, &t) >= 0)
+                free_and_replace(c->session, t);
+        else
+                c->session = mfree(c->session);
```
Comment (keszybz) [inline src/journal/journald-context.c]: Maybe it's too much magic, but lines 276–279 can be replaced with
```C
(void) cg_path_get_session(c->cgroup, &t);
free_and_replace(c->session, t);
```

File: src/journal/journald-context.c [hunk 24]
```diff
@@ -0,0 +1,600 @@
...
+int client_context_acquire(
+                Server *s,
+                pid_t pid,
+                const struct ucred *ucred,
+                const char *label, size_t label_len,
+                const char *unit_id,
+                ClientContext **ret) {
+
+        return client_context_get_internal(s, pid, ucred, label, label_len, unit_id, true, ret);
+};
```
Comment (keszybz) [inline src/journal/journald-context.c:535]: Shouldn't those be `static inline` functions? Or are we counting on lto to figure things out for us?

File: src/journal/journald-context.c [hunk 25]
```diff
@@ -0,0 +1,600 @@
...
+        int r;
+
+        assert(s);
+
+        /* Ensure that our own and PID1's contexts are always pinned. Our own context is particularly useful to
+         * generate driver messages. */
+
+        if (!s->my_context) {
+                struct ucred ucred = {
+                        .pid = getpid(),
```
Comment (keszybz) [inline src/journal/journald-context.c]: getpid_cached()?

File: src/journal/journald-server.c [hunk 26]
```diff
@@ -714,323 +715,109 @@ static void write_to_journal(Server *s, uid_t uid, struct iovec *iovec, unsigned
...
+                k = newa(char, strlen(field "=") + value_size + 1);     \
+                *((char*) mempcpy(stpcpy(k, field "="), value, value_size)) = 0; \
+                IOVEC_SET_STRING(iovec[n++], k);                        \
+        }                                                               \
 
 static void dispatch_message_real(
                 Server *s,
                 struct iovec *iovec, unsigned n, unsigned m,
-                const struct ucred *ucred,
+                ClientContext *c,
```
Comment (keszybz) [inline src/journal/journald-server.c]: `const`?

File: src/basic/process-util.c (hunk 17 above)
Comment (poettering) [reply src/basic/process-util.c]: yupp, this PR predates the getpid_cached() PR, hence it's based on a version without it. Will rebase.

File: src/basic/alloc-util.h (hunk 18 above)
Comment (poettering) [reply src/basic/alloc-util.h]: ouch!

File: src/journal/journald-context.c (hunk 20 above)
Comment (poettering) [reply src/journal/journald-context.c]: nope

File: src/journal/journald-server.c (hunk 21 above)
Comment (poettering) [reply src/journal/journald-server.c]: indeed!

File: src/journal/journald-context.c (hunk 22 above)
Comment (poettering) [reply src/journal/journald-context.c:258]: yes, it is... we trust the data from /proc more, and only use the data from the peer if we have nothing better. Hence we only set c->unit if it is NULL so far, as the comment is supposed to clarify.

File: src/journal/journald-context.c (hunk 24 above)
Comment (poettering) [reply src/journal/journald-context.c:535]: client_context_get_internal() is a static function, and gcc should be smart enough to optimize this away for us, and at least turn this into JMP rather than CALL, which I am very sure is good enough

**Comment by poettering** [comment]
force pushed a new version, please have a look. addressed all issues raised

File: src/journal/journald-context.c (hunk 22 above)
Comment (keszybz) [reply src/journal/journald-context.c:258]: Ah, right.

File: src/journal/journald-server.c [hunk 27]
```diff
@@ -1147,45 +926,21 @@ void server_dispatch_message(
...
-                c = strchr(c+1, '/');
-                if (c) {
-                        c = strchr(c+1, '/');
-                        if (c)
-                                *c = 0;
-                }
+                /* Write a suppression message if we suppressed something */
+                if (rl > 1)
+                        server_driver_message(s, "MESSAGE_ID=" SD_MESSAGE_JOURNAL_DROPPED_STR,
+                                              LOG_MESSAGE("Suppressed %u messages from %s", rl - 1, c->unit),
```
Comment (glasser) [inline src/journal/journald-server.c:939]: I actually have a script which parses this message and looks at the path (so eg, right now it is looking for `/system.slice/docker.service` here).  Can you confirm that after this change I should be looking for `docker.service` instead? That's the understanding I get from reading this PR but I don't actually know how to build and test systemd :)

Comment (poettering) [reply src/journal/journald-server.c:939]: i figure we should add _OBJECT_UNIT_NAME= to this line to make it easy to match against this log message in a structured way. But that should probably happen in a later commit

Comment (poettering) [reply src/journal/journald-server.c:939]: and yeah, you should check the unit name, not the cgroup path, as we do take liberty that the path might change

Comment (keszybz) [reply src/journal/journald-server.c:939]: Yes.

Comment (keszybz) [reply src/journal/journald-server.c:939]: It'd be nice to add SYSTEMD_UNIT= field too to that log message, so it shows up in 'journalctl -u' output, but that's not really related to this PR.

Comment (glasser) [reply src/journal/journald-server.c:939]: yeah, that would certainly be nicer than us having to run `journalctl -f _SYSTEMD_UNIT=docker.service + MESSAGE_ID=a596d6fe7bfa4994828e72309e95d61e` and then filter out based on parsing a string but I agree this is probably separate ;)

Comment (poettering) [reply src/journal/journald-server.c:939]: I filed #6494 as an RFE bug to add such a hook-up
